lightning/src/routing/gossip.rs

```diff
@@ -923,6 +923,47 @@ impl_writeable_tlv_based!(NodeAnnouncementInfo, {
 	(10, addresses, vec_type),
 });
 
+/// A user-defined name for a node, which may be used when displaying the node in a graph.
+///
+/// Node aliases provide a potential avenue for injection attacks. See [security considerations].
+///
+/// [security considerations]: https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#security-considerations-for-node-aliases
+#[derive(Clone, Debug, PartialEq)]
+pub struct NodeAlias([u8; 32]);
+
+impl fmt::Display for NodeAlias {
+	fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+		let control_symbol = core::char::REPLACEMENT_CHARACTER;
+		let first_null = self.0.iter().position(|b| *b == 0).unwrap_or(self.0.len());
+		let bytes = self.0.split_at(first_null).0;
+		let alias = match core::str::from_utf8(bytes) {
+			Ok(alias) => {
+				alias.chars()
+					.map(|c| if c.is_control() { control_symbol } else { c })
```

Comment:
Hmmm... the node in question isn't the user. Why should our users suffer for their bad behavior? If a node's alias contains a tab or newline, it would be surprising if we show a truncated or empty string, depending on its position.

I don't really understand the pushback here. This is pretty standard behavior and is a trivial amount of code. The alternative proposals are to truncate, as you initially suggested, or filter as you are now suggesting, which the only difference is swapping `map` for `filter` with a slightly simpler lambda. If you just don't like using a replacement character, that's fine, but I don't think the amount of code here is really a valid argument.

Reply:
> Right, I guess I don't see it as punishing the user. I see it as punishing the node who shoved garbage in their alias. Eg a tab may break someone printing node aliases in a tab-delimited file, so it's totally reasonable to just remove the crap the node tried to shove into the next column.
>

Let's say confusing the user then. 😉  e.g., When comparing to what a node explorer may show, a malicious node could be similarly aliased with but with some control characters thrown in. Probably no real security concern, but such a node could attempt to masquerade as another and would at very least be confusing.
 
> Yea, I think it's mostly all the iterators and allocating a string. Writing is usually slow/allocating anyway, but it's annoying to allocate just to filter. More generally, I find a bunch of iterators and chaining to be really hard to read, especially when there's a limit applied based on an index calculated elsewhere. I'm probably special here so not gonna hold this up anymore.

FWIW, the old `sanitize_string` from the sample was allocating a string. Not sure if we can get around that with `Formatter`'s interface. I guess we could write char-by-char using `char::encode_utf8`, though not sure how much more readable that would be.

Reply:
FWIW, without any allocations:

```diff
diff --git a/lightning/src/routing/gossip.rs b/lightning/src/routing/gossip.rs
index 0369a87c..ae2bd2c9 100644
--- a/lightning/src/routing/gossip.rs
+++ b/lightning/src/routing/gossip.rs
@@ -935,20 +935,30 @@ impl fmt::Display for NodeAlias {
                let control_symbol = core::char::REPLACEMENT_CHARACTER;
                let first_null = self.0.iter().position(|b| *b == 0).unwrap_or(self.0.len());
                let bytes = self.0.split_at(first_null).0;
-               let alias = match core::str::from_utf8(bytes) {
+               match core::str::from_utf8(bytes) {
                        Ok(alias) => {
-                               alias.chars()
-                                       .map(|c| if c.is_control() { control_symbol } else { c })
-                                       .collect::<String>()
+                               for c in alias.chars() {
+                                       let mut bytes = [0u8; 4];
+                                       if !c.is_control() {
+                                               f.write_str(c.encode_utf8(&mut bytes))?;
+                                       } else {
+                                               f.write_str(control_symbol.encode_utf8(&mut bytes))?;
+                                       }
+                               }
                        },
                        Err(_) => {
-                               bytes.iter().map(|b| *b as char)
+                               for c in bytes.iter().map(|b| *b as char) {
                                        // Display printable ASCII characters
-                                       .map(|c| if c >= '\x20' && c <= '\x7e' { c } else { control_symbol })
-                                       .collect::<String>()
+                                       let mut bytes = [0u8; 4];
+                                       if c >= '\x20' && c <= '\x7e' {
+                                               f.write_str(c.encode_utf8(&mut bytes))?;
+                                       } else {
+                                               f.write_str(control_symbol.encode_utf8(&mut bytes))?;
+                                       }
+                               }
                        },
                };
-               f.write_str(&alias)
+               Ok(())
        }
 }
```

Reply:
lol, true... guess that example doesn't make much sense. Pushed a version without allocations.